### PR TITLE
Implement Permit for token

### PR DESCRIPTION
**Pull request by MikitaS**
Implement permit function for sharded fungible token

### CODE REVIEW

File: ft-logic/src/lib.rs [hunk 1]
```diff
@@ -75,6 +75,35 @@ impl FTLogic {
                         self.approve(transaction_hash, account, &approved_account, amount)
                             .await;
                     }
+                    Action::PermitUnsigned {
```
Comment (LouiseMedova) [inline ft-logic/src/lib.rs:78]: What is this for?

File: ft-logic/src/lib.rs [hunk 2]
```diff
@@ -75,6 +75,35 @@ impl FTLogic {
...
+                        // TODO(m.strizhak) Where to report it?
+                        reply_err();
+                    }
+                    Action::Permit {
+                        owner_account,
+                        approved_account,
+                        amount,
+                        sign,
+                    } => {
+                        let payload = Action::PermitUnsigned {
```
Comment (LouiseMedova) [inline ft-logic/src/lib.rs:92]: `Action` has a meaning of the messages the `logic` contract can accept, so it's better to take `PermitInsigned` out.

File: ft-logic/src/lib.rs [hunk 3]
```diff
@@ -75,6 +75,35 @@ impl FTLogic {
...
                     }
+                    Action::PermitUnsigned {
+                        owner_account: _,
+                        approved_account: _,
+                        amount: _,
+                    } => {
+                        // TODO(m.strizhak) Where to report it?
+                        reply_err();
+                    }
+                    Action::Permit {
```
Comment (LouiseMedova) [inline ft-logic/src/lib.rs:86]: There is a problem with that signature. Anyone can send it as many times as they want after it has appeared on the chain. One of the possible solution is to add the time during which the signature is valid.

Comment (MikitaS) [reply ft-logic/src/lib.rs:78]: Solved. Protected with permit id

File: ft-logic/src/lib.rs (hunk 1 above)
Comment (MikitaS) [reply ft-logic/src/lib.rs]: struct was separated from Action

File: ft-logic/src/lib.rs (hunk 2 above)
Comment (MikitaS) [reply ft-logic/src/lib.rs]: Done

**Comment by MikitaS** [comment]
@LouiseMedova PTAL

**Review by LouiseMedova — CHANGES_REQUESTED** [review]

File: ft-logic/src/lib.rs [hunk 4]
```diff
@@ -259,6 +282,38 @@ impl FTLogic {
...
+    async fn permit(
+        &mut self,
+        transaction_hash: H256,
+        owner: &ActorId,
+        spender: &ActorId,
+        amount: u128,
+        owner_sign: &H512,
+        message: &PermitUnsigned,
+    ) {
+        if !self.check_permit_id(owner, &message.permit_id).await {
```
[the remark this answers is not in the document]
Comment (MikitaS) [reply ft-logic/src/lib.rs]: done

**Comment by MikitaS** [comment]
Rebased on latest master and locally checked

**Review by LouiseMedova — APPROVED** [review]

**Review by bezrazli4n0 — APPROVED** [review]

File: ft-storage/Cargo.toml [hunk 5]
```diff
@@ -4,7 +4,8 @@ version = "0.1.3"
 edition = "2021"
 
 [dependencies]
-gstd = { git = "https://github.com/gear-tech/gear.git", branch = "stable" }
+gstd = { git = "https://github.com/gear-tech/gear.git", branch = "stable", features = ["debug"]}
+light_sr25519 = { git = "https://github.com/gear-tech/gear.git", rev = "d23e90b6"}
```
Comment (fluiderson) [inline ft-storage/Cargo.toml:8]:
```suggestion
light_sr25519 = { git = "https://github.com/gear-tech/gear.git", branch = "stable" }
```

File: ft-main/tests/utils.rs [hunk 6]
```diff
@@ -1,11 +1,15 @@
...
+use sp_core::sr25519::Signature;
+
 pub trait FToken {
     fn ftoken(system: &System) -> Program;
     fn mint(&self, transaction_id: u64, from: u64, account: u64, amount: u128, error: bool);
     fn check_balance(&self, account: u64, expected_amount: u128);
+    fn check_balance_actor_id(&self, account: ActorId, expected_amount: u128);
+    fn check_permit_id(&self, account: ActorId, expected_permit_id: u128);
     fn burn(&self, transaction_id: u64, from: u64, account: u64, amount: u128, error: bool);
     fn transfer(
```
Comment (fluiderson) [inline ft-main/tests/utils.rs:14]:
```rs
    fn transfer(
        &self,
        transaction_id: u64,
        from: u64,
        sender: impl Into<ActorId>,
        recipient: impl Into<ActorId>,
        amount: u128,
        error: bool,
    );
```

File: ft-main/tests/utils.rs [hunk 7]
```diff
@@ -16,6 +20,15 @@ pub trait FToken {
...
     );
+    fn transfer_actor_id(
+        &self,
+        transaction_id: u64,
+        from: u64,
+        sender: ActorId,
+        recipient: ActorId,
+        amount: u128,
+        error: bool,
+    );
```
Comment (fluiderson) [inline ft-main/tests/utils.rs:31]:
```suggestion
```
See the previous comment.

File: ft-main/tests/utils.rs [hunk 8]
```diff
@@ -1,11 +1,15 @@
...
 use ft_main_io::*;
-use gstd::prelude::*;
+use gstd::{prelude::*, ActorId};
 use gtest::{Program, System};
+use sp_core::sr25519::Signature;
+
 pub trait FToken {
     fn ftoken(system: &System) -> Program;
     fn mint(&self, transaction_id: u64, from: u64, account: u64, amount: u128, error: bool);
     fn check_balance(&self, account: u64, expected_amount: u128);
```
Comment (fluiderson) [inline ft-main/tests/utils.rs:10]:
```rs
    fn check_balance(&self, account: impl Into<ActorId>, expected_amount: u128);
```

File: ft-main/tests/utils.rs [hunk 9]
```diff
@@ -1,11 +1,15 @@
...
+    fn check_balance_actor_id(&self, account: ActorId, expected_amount: u128);
```
Comment (fluiderson) [inline ft-main/tests/utils.rs:11]:
```suggestion
```
See the previous comment.

File: ft-main/tests/utils.rs [hunk 10]
```diff
@@ -1,11 +1,15 @@
...
+    fn check_permit_id(&self, account: ActorId, expected_permit_id: u128);
```
Comment (fluiderson) [inline ft-main/tests/utils.rs:12]:
```suggestion
    fn check_permit_id(&self, account: [u8; 32], expected_permit_id: u128);
```

File: ft-main/tests/utils.rs [hunk 11]
```diff
@@ -129,10 +182,51 @@ impl FToken for Program<'_> {
...
+    }
+
+    fn check_balance_actor_id(&self, account: ActorId, expected_amount: u128) {
+        let res = self.send(HARDCODED_ACCOUNT, FTokenAction::GetBalance(account));
+        let reply = FTLogicEvent::Balance(expected_amount).encode();
+        assert!(res.contains(&(HARDCODED_ACCOUNT, reply)));
+    }
+
+    fn check_permit_id(&self, account: ActorId, expected_permit_id: u128) {
+        let res = self.send(HARDCODED_ACCOUNT, FTokenAction::GetPermitId(account));
```
Comment (fluiderson) [inline ft-main/tests/utils.rs:227]:
```suggestion
    fn check_permit_id(&self, account: [u8; 32], expected_permit_id: u128) {
        let res = self.send(HARDCODED_ACCOUNT, FTokenAction::GetPermitId(account.into()));
```

**Review by fluiderson — APPROVED** [review]

File: ft-logic/Cargo.toml [hunk 12]
```diff
@@ -4,11 +4,12 @@ version = "0.1.3"
 edition = "2021"
 
 [dependencies]
-gstd = { git = "https://github.com/gear-tech/gear.git", branch = "stable" }
+gstd = { git = "https://github.com/gear-tech/gear.git", branch = "stable", features = ["debug"]}
 ft-logic-io = { path = "io" }
 ft-storage-io = { path = "../ft-storage/io" }
 primitive-types = { version = "0.12.1", default-features = false, features = ["scale-info"]}
 hex = { version = "0.4.3", default-features = false, features = ["alloc"] }
+light_sr25519 = { git = "https://github.com/gear-tech/gear.git", rev = "d23e90b6"}
```
[the remark this answers is not in the document]
Comment (MikitaS) [reply ft-logic/Cargo.toml:12]: done

File: ft-storage/Cargo.toml (hunk 5 above)
Comment (MikitaS) [reply ft-storage/Cargo.toml:8]: done

File: ft-main/tests/utils.rs (hunk 11 above)
Comment (MikitaS) [reply ft-main/tests/utils.rs:227]: done

File: ft-main/tests/utils.rs (hunk 10 above)
Comment (MikitaS) [reply ft-main/tests/utils.rs:12]: done

File: ft-main/Cargo.toml [hunk 13]
```diff
@@ -16,3 +16,8 @@ gear-wasm-builder = { git = "https://github.com/gear-tech/gear.git", branch = "s
 
 [dev-dependencies]
 gtest = { git = "https://github.com/gear-tech/gear.git", branch = "stable" }
+hex-literal = "0.3.4"
+sp-core = { version = "6.0.0", git = "https://github.com/gear-tech/substrate.git", branch = "gear-stable" }
+light_sr25519 = { git = "https://github.com/gear-tech/gear.git", rev = "d23e90b6" }
```
[the remark this answers is not in the document]
Comment (MikitaS) [reply ft-main/Cargo.toml:21]: done

File: ft-main/tests/utils.rs (hunk 8 above)
Comment (MikitaS) [reply ft-main/tests/utils.rs:10]: done

File: ft-main/tests/utils.rs (hunk 7 above)
Comment (MikitaS) [reply ft-main/tests/utils.rs:31]: done

File: ft-main/tests/utils.rs (hunk 9 above)
Comment (MikitaS) [reply ft-main/tests/utils.rs:11]: done

File: ft-main/tests/utils.rs (hunk 6 above)
Comment (MikitaS) [reply ft-main/tests/utils.rs:14]: done

**Comment by MikitaS** [comment]
All comments are fixed

**Pull request merged**